### PR TITLE
Refactor scheduler data structures

### PR DESCRIPTION
**Pull request by raulchen**
## Why are these changes needed?

This is the first PR to refactor scheduler data structures (See #22850).

Major changes:
- Hid the implementation details in the `ResourceRequest` and `TaskResourceInstnaces` classes, which expose public methods such as algebra operators and comparison operators. 
- Hid the differences between "predefined" and "custom" resources inside these 2 classes. Call sites can simply use the resource ID to access the resource, no matter it is predefined or custom.
- The predefined_resources vector now always has the full length. So no more "resize"s are needed. 
- Removed the `ResourceCapacity` class. Now "total" and "available" resources are stored in separate fields in "NodeResources". 
- Moved helper functions for FixedPoint vectors from "cluster_resource_data.h" to "fixed_point.h"
- "ResourceID" now has static methods to get the resource ids of predefined resources, e.g. "ResourceID::CPU()". 
- Encapsulated unit-instance resource logic to "ResourceID"

Other planned changes that are not included in this PR:
- Rename ResourceRequest to ResourceSet, and move it to its own file.
- Remove the predefined vectors and always use maps.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (scv119) [inline src/ray/raylet/scheduling/cluster_resource_data.h:145]: hmm this function call involves constructing a set. we should benchmark the scheduling performance...
maybe we should really remove the vector optimization.

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (scv119) [inline src/ray/raylet/scheduling/cluster_resource_data.h:139]: let's make this O(1)?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (scv119) [inline src/ray/raylet/scheduling/cluster_resource_data.h:403]: same here; we should make this O(1)

**Comment by scv119** [comment]
Thanks, it looks pretty good!  @rkooo567  @iycheng  take another pass?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_data.h:153]: Yeah, I'll do a micro benchmark.

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_data.h]: I can't make it fully O(1). Because we need to filter out the 0-value resources. 
I've make the "custom_resources" to remove 0-value resources automatically. But it's not possible for predefined_resources. 
After the latest optimization, the complexity is now O(len(predefined_resources) + 1).

**Comment by raulchen** [comment]
@scv119 @iycheng @rkooo567 @wumuzi520 @Chong-Li @WangTaoTheTonic 
Hi everyone. This PR is ready for review. 
The only remaining work is to fix "cluster_resource_scheduler_test", which is too complex and I need to discuss with @scv119 offline. 
You can take a look at other files first.

File: src/ray/common/bundle_spec.cc
Comment (WangTaoTheTonic) [inline src/ray/common/bundle_spec.cc:39]: `auto &resoure_id` ?

File: src/ray/common/bundle_spec.cc
Comment (WangTaoTheTonic) [inline src/ray/common/bundle_spec.cc:41]: Here we can compute the `ToDouble` of `resource_value` to avoid extra once compute later.

**Review by WangTaoTheTonic — COMMENTED** [review]
A lot of `const auto &resource_name = resource_id.Binary();` are found in codes.
Even I know that `resource_id` is not a simple literal id but a `ResourceID` object, I still need a tiny switch in my head while I read that line, because `id.binary -> name` is kinda strange.
Usually the logic is `name.binary -> id`, is that right or only my personal feel?
(I know it is not caused by this PR, but I wanna bring it up to discuss)
also @scv119

File: src/ray/gcs/gcs_server/gcs_resource_scheduler.cc
Comment (WangTaoTheTonic) [inline src/ray/gcs/gcs_server/gcs_resource_scheduler.cc:40]: &resource_id

File: src/ray/gcs/gcs_server/gcs_resource_scheduler.cc
Comment (WangTaoTheTonic) [inline src/ray/gcs/gcs_server/gcs_resource_scheduler.cc:283]: &resource_id

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (WangTaoTheTonic) [inline src/ray/raylet/scheduling/cluster_resource_data.h:101]: Contains?

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (WangTaoTheTonic) [inline src/ray/raylet/scheduling/cluster_resource_data.cc:98]: &resource_id

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (WangTaoTheTonic) [inline src/ray/raylet/scheduling/cluster_resource_data.cc:138]: Maybe we could rename to `Contains` to be consistent with the codes elsewhere.

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (WangTaoTheTonic) [inline src/ray/raylet/scheduling/cluster_resource_data.h:141]: we can use a `std::count` to simplify the codes?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (WangTaoTheTonic) [inline src/ray/raylet/scheduling/cluster_resource_data.h:178]: move this to first line to short cut the comparison?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (WangTaoTheTonic) [inline src/ray/raylet/scheduling/cluster_resource_data.h:228]: Should we just use `IsSubset` instead of `<=`?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (WangTaoTheTonic) [inline src/ray/raylet/scheduling/cluster_resource_data.h:355]: `Contains`?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (WangTaoTheTonic) [inline src/ray/raylet/scheduling/cluster_resource_data.h:488]: Here we can iterate all resource ids to remove this TODO?

File: src/ray/common/bundle_spec.cc
Comment (raulchen) [reply src/ray/common/bundle_spec.cc:41]: IMO, we should always use FixedPoint. Because FixedPoint is meant supposed to deal with the precision issues of doubles. The "Double()" method shouldn't exist at all. But in this PR, I just keep them as they are.

File: src/ray/common/bundle_spec.cc
Comment (raulchen) [reply src/ray/common/bundle_spec.cc:39]: I didn't use references because the resource id is just an integer. 
Although this is not likely to change, I can still add references as it doesn't hurt any way.

**Comment by raulchen** [comment]
> A lot of `const auto &resource_name = resource_id.Binary();` are found in codes.
> Even I know that `resource_id` is not a simple literal id but a `ResourceID` object, I still need a tiny switch in my head while I read that line, because `id.binary -> name` is kinda strange.
> Usually the logic is `name.binary -> id`, is that right or only my personal feel?
> (I know it is not caused by this PR, but I wanna bring it up to discuss)
> also @scv119

I think it's better to rename "Binary" to something like "Name" or "Label". @WangTaoTheTonic @scv119  
I can do it in a different PR, if you all agree.

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_data.h:488]: I'll leave it to @chenk008 
For this PR, I'd like to only do refactor and minimize behavior changes.

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_data.cc:138]: Searched a bit. "Contains" doesn't seem to be widely used either. 
Considering "Has" is also pretty clear. I'm in favor of keeping "Has".

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_data.h:228]: I thought of "IsSubset". But changed to "<=" as "a <= b" looks clearer to me. No super strong preference thought.

File: src/ray/gcs/gcs_server/gcs_resource_scheduler.cc
Comment (WangTaoTheTonic) [inline src/ray/gcs/gcs_server/gcs_resource_scheduler.cc:35]: Is it better to `RemoveNegative` in `-=` implicitly? It's more friendly for forgetful coders.

Comment (raulchen) [reply src/ray/gcs/gcs_server/gcs_resource_scheduler.cc:35]: Negative values are allowed in "ResourceRequest". For example, you may want to calculate the diff of 2 sets. So we cannot do that by default.

**Comment by rkooo567** [comment]
Is this ready to review?

**Comment by raulchen** [comment]
@rkooo567 yeah, I've said that in a previous comment. Actually, I think it should be ready for merge.

**Comment by raulchen** [comment]
Test failures should be irrelevant.

**Comment by scv119** [comment]
@iycheng @jjyao this pr hugely simplify our resource related code. can you take a look? I have already given it a go and it looks great.

**Comment by jjyao** [comment]
Love more reds than greens! I'll take a look tomorrow.

**Comment by raulchen** [comment]
@jjyao @iycheng To simplify review, the major meaningful changes are in the "src/ray/raylet/scheduling/cluster_resource_data.h" file. Other files are just changing the usages.

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_data.h:65]: nit: 0 -> nullptr

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_data.h:53]: I think we can just do `: predefined_resources_(PredefinedResourcesEnum_MAX), requires_object_store_memory_(requires_object_store_memory) {`?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (jjyao) [reply src/ray/raylet/scheduling/cluster_resource_data.h:136]: If perf is not an issue, we can just do `ResourceIds().size()`?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_data.h:170]: This is auto generated by compiler?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_data.h:196]: What about the else case?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_data.h:215]: What if negative?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_data.h:230]: Could we update the comment to reflect that the value is also compared, not just one set has more resource types than the other.

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_data.h:242]: What about this case `it == other.custom_resources_.end() && entry.second < 0`?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_data.h:261]: Binary() is not going to be readable I think, maybe add a `Hex()` method?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_data.h:375]: Do we also need to check sum(instances) is 0 as well?

File: src/ray/raylet/scheduling/fixed_point.cc
Comment (jjyao) [inline src/ray/raylet/scheduling/fixed_point.cc:19]: Won't `v1 == v2` just work?

File: src/ray/raylet/scheduling/scheduling_policy_test.cc
Comment (jjyao) [inline src/ray/raylet/scheduling/scheduling_policy_test.cc:367]: Is this moved somewhere else?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (scv119) [inline src/ray/raylet/scheduling/cluster_resource_data.h:39]: looks we need a bit more unit test to ensure the edge cases are covered

**Comment by scv119** [comment]
for change like this, let's at least get 3 approvals.

File: src/ray/raylet/scheduling/scheduling_policy_test.cc
Comment (raulchen) [reply src/ray/raylet/scheduling/scheduling_policy_test.cc:367]: It's moved to another file in a previous PR by @scv119. He probably forgot to remove this.

File: src/ray/raylet/scheduling/fixed_point.cc
Comment (raulchen) [reply src/ray/raylet/scheduling/fixed_point.cc:19]: Yeah, I just moved it from another file. Will remove it.

**Comment by raulchen** [comment]
@jjyao Thanks for your comments. All problems should have been addressed.

**Review by scv119 — APPROVED** [review]

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_data.h:206]: I feel we expect `this` and `other` have the same set of resources? I'm not sure if it's a sensible to do things like `ResourceRequest(A, B, C) -/+ ResourceRequest(D, E, F)`

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_data.h:396]: Can this ever happen?

**Review by jjyao — COMMENTED** [review]
LG. My main question is about removing custom resource when it's 0. Is this needed since we don't/can't remove predefined resources when it's 0 anyway.

**Comment by raulchen** [comment]
> LG. My main question is about removing custom resource when it's 0. Is this needed since we don't/can't remove predefined resources when it's 0 anyway.

@jjyao The problem is that previously, when we don't remove 0 resources, all the use sites have to repeatedly check if the value is 0. I think this is an unnecessary burden for the users. They shouldn't care wether "a resource set has a particular resource with 0 value" or "a resource set doesn't have this particular resource". Semantically, these 2 cases should be retreated as the same thing. 

As for removing the entry in the custom resources map, the reason is that by doing so, we can simply use "custom_resources_.size()` without needing to filtering 0 values.

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_data.h:90]: Nope. the current API discourages this. See my comment on the main PR page for details.

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_data.h:206]: I think it's valid. Because we do dynamically adding new resources (PG resources for example) to node.

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_data.h:396]: I'm not sure if we already have such a use case. But from this API's perspective, it's allowed. It's equivalent to "ResourceRequest::Set(resource_id, 0)".

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_data.h:36]: Replied here https://github.com/ray-project/ray/pull/22854#issuecomment-1080296039

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_data.h:109]: That makes sense. But in practice, the 2 sets can have different resources. Because the value in the available can be reduced to 0. 
Also, I think this method should be removed. If even "available > total", we can still keep that. I didn't want to change the behavior and then added this "Cap" method. I can do some investigation afterwards. 
For this PR, I'll make this method generic.

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_data.h:168]: Are you suggesting "= default"?

Comment (jjyao) [reply src/ray/raylet/scheduling/cluster_resource_data.h]: You don't even need "=default". We can just omit this function and compiler will generate one for us I think. But "=default" is also fine to me.

File: src/ray/gcs/gcs_server/gcs_resource_manager.cc
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_resource_manager.cc:134]: `available` can be negative and the old code filters it out. We should do the same thing here?

File: src/ray/raylet/scheduling/cluster_resource_manager.cc
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_manager.cc:202]: Can we check to make sure resource_request doesn't contain custom resources that node_resources doesn't have? Otherwise I think the old code and the new one don't have the same behavior.

**Review by jjyao — APPROVED** [review]

File: src/ray/common/bundle_spec.cc
Comment (raulchen) [reply src/ray/common/bundle_spec.cc:39]: I think most of the "<= 0" filters were simplify copy-pasted. They should be unnecessary. 
I'd like to keep remove these filters unless we really understand why it is needed.
Considering all tests have passed, I think we can keep the current code logic.

File: src/ray/raylet/scheduling/cluster_resource_manager.cc
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_manager.cc:202]: Good catch. 

@scv119 what do you think is the correct semantics of this function? Should it only add existing resources?

There is another "UpdateNodeAvailableResourcesIfExist" function that explicitly says it only update existing resources. But this "AddNodeAvailableResources" function doesn't.

Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_manager.cc]: I'll keep the behavior unchanged in this PR (only add resources that only exist in node_resources).

Comment (scv119) [reply src/ray/raylet/scheduling/cluster_resource_manager.cc]: yeah better keep the current behavior... probably we should merge it with UpdateNodeAvailableResourcesIfExist in follow up PRs.

**Comment by raulchen** [comment]
@jjyao @scv119 thanks for your comments. All problems should be addressed. 
I'll merge this PR when tests pass, in order to unblock other PRs. Please let me know if you have other questions.

**Comment by raulchen** [comment]
Tests failures should be unrelated.

**Pull request merged**
